### PR TITLE
Adds docs on multi-controller JAX fault tolerance.

### PR DESCRIPTION
**Pull request by mwhittaker**

### CODE REVIEW

File: docs/fault_tolerance.rst [hunk 1]
```diff
@@ -0,0 +1,1522 @@
...
+    else:
+      ... # Branch B
+    print(y)
+
+See :ref:`part3` for details on how atomicity is implemented.
+
+Sharding Arrays
+^^^^^^^^^^^^^^^
+
+TODO: Double check if ``jax.device_put`` works when some machines have
```
Comment (samos123) [inline docs/fault_tolerance.rst]: Not sure if you want to address this before merging. I'm curious.

File: docs/_static/fault_tolerance/data_parallelism.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  return jax.make_array_from_single_device_arrays(x.shape, sharding, shards)
+
+
+def main(_: Sequence[str]) -> None:
+  # Parse command line arguments and initialize multi-controller JAX.
+  jax.config.update("jax_enable_recoverability", True)
+  jax.distributed.initialize(coordinator_address="localhost:8000",
+                             process_id=_PROCESS_ID.value,
+                             num_processes=_NUM_PROCESSES.value,
+                             local_device_ids=[process_id],
```
Comment (samos123) [inline docs/_static/fault_tolerance/data_parallelism.py]: this will cause an error since process_id isn't defined.

```suggestion
                             local_device_ids=[_PROCESS_ID.value],
```

File: docs/_static/fault_tolerance/data_parallelism_with_recovery.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  return jax.block_until_ready(f(np.array([x] * len(devices)))).addressable_shards[0].data
+
+
+def main(_: Sequence[str]) -> None:
+  # Parse command line arguments and initialize multi-controller JAX.
+  jax.config.update("jax_enable_recoverability", True)
+  jax.distributed.initialize(coordinator_address="localhost:8000",
+                             process_id=_PROCESS_ID.value,
+                             num_processes=_NUM_PROCESSES.value,
+                             local_device_ids=[process_id],
```
Comment (samos123) [inline docs/_static/fault_tolerance/data_parallelism_with_recovery.py]: same issue

```suggestion
                             local_device_ids=[_PROCESS_ID.value],
```

Comment (mwhittaker) [reply docs/_static/fault_tolerance/data_parallelism_with_recovery.py]: Done.

File: docs/_static/fault_tolerance/data_parallelism.py (hunk 2 above)
Comment (mwhittaker) [reply docs/_static/fault_tolerance/data_parallelism.py]: Done.

File: docs/fault_tolerance.rst (hunk 1 above)
Comment (mwhittaker) [reply docs/fault_tolerance.rst]: I'll leave it for later.

File: docs/fault_tolerance.rst [hunk 4]
```diff
@@ -0,0 +1,1516 @@
...
+Again run the script across four processes and then kill the fourth. Notice
+that now, the other three processes happily continue executing.
+
+Next try failing process 0. Notice that all four processes terminate with
+error messages that look something like the following:
+
+.. code-block::
+
+   E0929 17:42:48.594192 1044529 coordination_service_agent.cc:332] Polled an error from coordination service (this can be an error from this or another task).
+   F0929 17:42:48.594200 1044529 client.h:77] Terminating process because the JAX distributed service detected fatal errors. This most likely indicates that another task died; see the other task logs for more details. Disable Python buffering, i.e. `python -u`, to be sure to see all the previous output. absl::Status: UNAVAILABLE: Failed to send RPC to coordination service. Either the leader task was preempted/died/restarted unexpectedly or this task is experiencing network issues. Check earlier logs from 1) this task, 2) the leader (usually slice 0 task 0), and 3) cluster scheduler to debug further.
```
Comment (mattjj) [inline docs/fault_tolerance.rst:108]: Tangential: These error messages kinda suck. They don't foreground the most important info, and they're all C++y. Everyone who hits this is going to bug Yash in chat. Can we improve them?

The important stuff might be "Check earlier logs from 1) this task, 2) the leader (usually slice 0 task 0), and 3) cluster scheduler to debug further."

File: docs/_static/fault_tolerance/collectives.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      local_device_ids=[_PROCESS_ID.value],
+      heartbeat_timeout_seconds=10,
+  )
+  print(f'{jax.devices()=}')
+  print(f'{jax.local_devices()=}')
+  while True:
+    mesh = jax.make_mesh((4,), ("i",))
+    spec = jax.sharding.PartitionSpec("i")
+    sharding = jax.sharding.NamedSharding(mesh, spec)
+    x = jax.device_put(jnp.array([1., 2., 3., 4.]), sharding)
```
Comment (mattjj) [inline docs/_static/fault_tolerance/collectives.py]:
```suggestion
    jax.set_mesh(jax.make_mesh((4,), ("i",)))    # move outside the while_loop?
    x = jax.device_put(jnp.array([1., 2., 3., 4.]), jax.P("i"))
```

Comment (mattjj) [reply docs/_static/fault_tolerance/collectives.py]: Ohhh, I see why you put it in the loop; it's for the `with live_devices(...) as devices` stuff later...

File: docs/fault_tolerance.rst (hunk 4 above)
Comment (mwhittaker) [reply docs/fault_tolerance.rst:108]: Good point. Let me improve the error messages in a future change.

File: docs/_static/fault_tolerance/collectives.py (hunk 5 above)
Comment (mwhittaker) [reply docs/_static/fault_tolerance/collectives.py]: Updated the code and moved the mesh stuff outside the while loop.

File: docs/fault_tolerance.rst [hunk 6]
```diff
@@ -0,0 +1,1516 @@
...
+
+.. literalinclude:: _static/fault_tolerance/data_parallelism.py
+    :language: python
+    :lines: 1-22
+    :lineno-start: 1
+
+Next, we define a ``replicated`` function that returns an array replicated
+across a set of devices. Note that ``replicated`` doesn't actually replicate
+any data. It assumes the data is already replicated. It simply returns a
+``replicated`` view of the data.
```
Comment (mattjj) [inline docs/fault_tolerance.rst]:
```suggestion
across a set of devices. Note that ``replicated`` doesn't actually move
any data. It assumes the argument ``x`` already has equal value across all processes. It simply returns a new view of that data, in a process-spanning `jax.Array` with a replicated sharding.
```

Comment (mattjj) [reply docs/fault_tolerance.rst]: My thinking was that the previous text was using the term "replicated" a bit circularly.

**Review by mattjj — APPROVED** [review]
Brilliant stuff, thank you! What I love most is that by reading this I learned about distributed fault tolerance fundamentals, not just parochial JAX APIs.

Let's get this in, and we can continue to iterate if we see little wording tweaks and stuff.

File: docs/fault_tolerance.rst (hunk 6 above)
Comment (mwhittaker) [reply docs/fault_tolerance.rst]: Thanks Matt, your wording is much clearer.

**Comment by mwhittaker** [comment]
> Brilliant stuff, thank you! What I love most is that by reading this I learned about distributed fault tolerance fundamentals, not just parochial JAX APIs.
> 
> Let's get this in, and we can continue to iterate if we see little wording tweaks and stuff.

Thank you! And yes, I totally agree. I also want to make some changes to include some text or maybe a screen recording of the `example.py` programs running to make the tutorial easier to follow without having to run the scripts yourself. And I want to add some text on how you can play with these features on CPU instead of GPU, which I'm guessing is easier for most people.
